pep-0728.rst

```diff
@@ -0,0 +1,322 @@
+
+.. code-block:: py
+
+    # This definition is obviously simplified for the sake of the example.
+    type Filter[T: dict, TName: FieldName[T]] = dict[TName, NotRequired[Field[T, TName]]]
+
+    class Collection[T]:
+        def find(self, filter: Filter[T]) -> Iterable[T]:
+            ...
+
+    cat_collection: Collection[Cat]
+
+    # Passes type checking because filters do not require all of Cat's fields to be present within the query.
+    cat_collection.find({"age": 3})
+
+    # Fails type checking because fur_color must be a string
+    cat_collection.find({"fur_color": 1})
+
+    # Fails type checking because "foo" is not declared in the collection's scheme
+    cat_collection.find({"foo": "bar"})
+
+
+Specification
+=============
+
+The current proposals introduces adds ``Field`` and ``FieldName`` to the ``typing`` module.
+
+Field
+'''''''''
+``Field`` can be used to refer to another type's fields. It receives 2 arguments, a type and a name of a field within
+this type. The first argument may be a dataclasses, (or dataclass-transformed objects), classes implementing
+``__slots__``, ``TypedDict`` or any other type with a typed layout.
+
+.. code-block:: py
+
+    from typing import Field
+
+    @dataclass
+    class Foo:
+        bar: int
+        baz: str
+
+    # Function returns int
+    def func() -> Field[Foo, 'bar']:
+        pass
+
+While dataclass-like objects will consider the attributes of the instance, dictionaries will instead refer to the items
+of the instance.
+
+.. code-block:: py
+
+    from typing import Field, TypedDict
+
+    class Foo(TypedDict):
+        bar: int
+        baz: str
+
+    # Function returns int
+    def func() -> Field[Foo, 'bar']:
+        pass
+
+
+Multiple field names may be specified using ``Literal``. In these cases, the annotation will be the
+equivalent of a union of the types of the given fields.
+
+.. code-block:: py
+
+    from typing import Field, Literal
+
+    @dataclass
+    class Foo:
+        bar: int
+        baz: str
+
+    # Function returns int | str
+    def func() -> Field[Foo, Literal['bar', 'baz']]:
+        pass
+
+Using ``Field`` with an explicit ``Literal`` of a single field is allowed, but not needed because you can use the bare
+field name.
+
+.. code-block:: py
+
+    from typing import Field, Literal
+
+    @dataclass
+    class Foo:
+        bar: int
+        baz: str
+
+    # Function returns int
+    def func() -> Field[Foo, Literal["bar"]]:
+        pass
+
+
+String field names that are provided to ``Field`` will always be assumed to be the literal strings (instead of forward
+references) unless the entire annotation is wrapped within quotes. When the entire annotation is wrapped within quotes,
+you may still explicitly use ``Literal`` in order to prevent this behavior.
+
+.. code-block:: py
+
+    from typing import Field, Literal
+
+    @dataclass
+    class Foo:
+        bar: int
+        baz: str
+
+    bar: TypeAlias = Literal['baz']
```

Comment:
Should this use the type parameter syntax since TypeAlias is deprecated(-ish)? They technically don't do the same thing at runtime but I'm guessing both are supposed to work the same within the context of this PEP's new typing constructs.